carbon-design-system/carbon
# feat(data-table): adds experimental styles and new functionalities

Closes #1303 

A lot of changes so far. Still a WIP, but adding this PR now so that I can start refactoring anything based on comments. My biggest concern would be the changes I made for tooltip functionality and expansion.

#### Changelog
**New Specs for reference**
- numerical columns can be right aligned (optional)
- sticky table header (optional)
- overflow-x scroll only on table (optional)
-  text truncation/tooltip (optional)
- hidden search bar (default)
- persistent search (optional)

**New Changes**
- adds `spans` to `td` and `th` to allow for text truncation 
- adds tooltip div to table cells
- adds optional header text
- adds hidden search bar and persistent search bar
- adds js for tooltip functionality
- adds event listener on window resizing to determine text truncation changes
- adds conditional table container for optional horizontal scroll (meaning the toolbar will not scroll with the table on overflow-x) 
- new classes:

`.#{$prefix}--data-table-v2--numerical-cell `   will right align text in `td` for numerical columns 
`.#{$prefix}--data-table-v2--fixed-header `    will add styles for sticky header 
`.#{$prefix}--data-table-v2--truncated`   will add tooltip visibility on hover 
`.#{$prefix}--data-table-v2--overflow-truncate`  will add text truncation to data table 
`.#{$prefix}--data-table-v2--overflow-scroll`  will scroll only the table and not the toolbar/header on horizontal overflow
`.#{$prefix}--data-table-v2--fixed-toolbar` will scroll only the table and not the toolbar/header on horizontal overflow
`.#{$prefix}--table-header-label`  span tag for `th` text
`.#{$prefix}--table-cell-content`  span tag for `td` text
`.#{$prefix}--table-column-checkbox` specifies width for checkbox table cells (width is diff for other table cells)
`.#{$prefix}--table-column-menu ` specifies width for overflow menu table cells (width is diff for other table cells)

**Changed**
- overall styles
- toolbar
- icons
- batch action
- updated row expansion to utilize more css instead of js (still uses some js but not as much; this is for both stable and experimental - mainly to allow for transitions/motion)

**Removed**
- commented out some js for row expansion (will delete all together but didn't wanna remove yet before others have tested/reviewed)

#### Testing / Reviewing
- check that Stable and Experimental data table expansion is working correctly
- check that tooltip functionality is being updated on window resizing (when text truncation is being toggled at different sizes)

#### To do
- [x] fix sticky header styles 
- [x] update pagination styles
- [x] add additional icon options for rows (to the left of overflow menu)
- [x] update config file for new examples
- [x] update tests to reflect js changes
